### ISSUE TITLE
When using Spark Thrift server to use Spark SQL，an exception occurs——File does not exist

### ISSUE THREAD

**Issue by ChenXiongfeng** [delta-io/delta#1470]
After using Spark SQL to transfer data to hdfs，I get an error when I query the contents of this table

`Error operating EXECUTE_STATEMENT: org.apache.spark.SparkException: Job aborted due to stage failure: Task 4 in stage 197.0 failed 4 times, most recent failure: Lost task 4.3 in stage 197.0 (TID 3521) (data198 executor 2): java.io.FileNotFoundException: File does not exist: hdfs://data197:8020/user/hive/warehouse/zilaishui_dwd.db/jbyjslmx1/part-00019-f2606e9e-691a-4188-ae6b-47d34ecaf37a-c000.snappy.parquet
It is possible the underlying files have been updated. You can explicitly invalidate the cache in Spark by running 'REFRESH TABLE tableName' command in SQL or by recreating the Dataset/DataFrame involved.`

The delta log does record the above file information, but the hdfs does not exist.The probability of this problem is very small.I don't know whether the problem is on HDFS or Delta, or my hardware.

Now the data in this table cannot be queried. How can I solve this problem

### Environment information

* Delta Lake version:1.0.0
* Spark version: 3.1.2
* Scala version: 2.12

**Comment by scottsand-db** [comment]
1. How did you transfer data to HDFS?
2. "The probability of this problem is very small" Do you consistently have this issue? Or is it random?
3. How are you loading your delta table? Which exact command is giving you this error?
4. Have you run VACUUM on your delta table?

**Comment by ChenXiongfeng** [comment]
1. Execute some associated and merge operation through Spark SQL（All operations use Delta）:

**1) Create temporary table**

```
create table dwd1020.tmp_5084664359800537220
(
    id                    string comment 'id',
    value_string_255      string comment 'value',
    state                 string comment 'state',
    name                  string comment 'name',
    insert_time           timestamp comment 'insert_time',
    quality_filter_result string comment 'quality_filter_result'
) USING DELTA;

```

**2).Associate multiple tables and insert temporary tables**
```
insert into `dwd1020`.tmp_5084664359800537220 (id, value_string_255, state, name, insert_time, quality_filter_result)
select tmp.id    as id,
       tmp.value as value_string_255,
       tmp.state as state,
       null                      as name,
       current_timestamp()     as insert_time,
       'UN_CHECK'                as quality_filter_result
from (select t_1582975638421008385.state                 as state,
             t_1582975647835230209.value_string_255      as value,
             t_1582976225164754946.id                    as id,
             t_1582976225164754946.insert_time           as insert_time,
             t_1582976225164754946.quality_filter_result as quality_filter_result
      from `ods1020`.ods1020_1 t_1582976225164754946
               left join `ods1020`.ods1020_2 t_1582975638421008385
                         on t_1582976225164754946.id = t_1582975638421008385.id
               left join `ods1020`.ods1020_3 t_1582975647835230209
                         on t_1582976225164754946.id = t_1582975647835230209.id
      where 1 = 1) tmp
where 1 = 1
  and tmp.insert_time >= to_timestamp('2022-10-21 18:00:00', 'yyyy-MM-dd HH:mm:ss')
  and tmp.insert_time < to_timestamp('2022-10-21 21:00:00', 'yyyy-MM-dd HH:mm:ss');
```

**3).Merge temporary table to target table**
```
merge into dwd1020.dwd1020_2 as t using dwd1020.tmp_5084664359800537220 as s on t.id = s.id
when matched and (t.id != s.id or t.value_string_255 != s.value_string_255 or t.state != s.state or t.name != s.name or
                  t.insert_time != s.insert_time or t.quality_filter_result !=
                                                    s.quality_filter_result) then update set id = s.id,value_string_255 = s.value_string_255,state = s.state,name = s.name,insert_time = s.insert_time,quality_filter_result = s.quality_filter_result
when not matched then insert *; 
```

3. Random. The data flow of our data platform is realized through Spark SQL, but suddenly this problem occurs when a table is queried one day, and the data flow before this table is normal.
4. It is a common query: select * from table.  
5. yes. We will execute this operation regularly and set to vacuum (0) .I know this statement will clean up the data of the old version, but keep the latest version.We do not specify a version for the query data, and use the latest version by default.

We solved the previous problem by rerunning sql task，now we're trying to recurri this problem... but everything is normal

**Comment by guoliqiang** [comment]
> 1. How did you transfer data to HDFS?
> 2. "The probability of this problem is very small" Do you consistently have this issue? Or is it random?
> 3. How are you loading your delta table? Which exact command is giving you this error?
> 4. Have you run VACUUM on your delta table?

Hi @scottsand-db I faced a same issue, I am using ADLS Gen2 (delta version 0.8), and my case is streaming case
1. Job 1 append data into delta table
2. Job 2 streaming read from delta table.

Job 2 encoutered this error (Job 2 find a file add entry from log, but faild to read that parquet file due to it does not exists), I manully checked the files, it indeed does not exists.

I also have a purge job (deletedFileRetentionDuration is 0 in my delta table) which will running in parallel.  The purge job run a delete command first to remove the data genrated 7 days ago, then run vacuum command.
 
I checked the log, I found the missed parquet file commit just followed the vacumm commit.  Do you have any insight?

**Comment by zsxwing** [comment]
> yes. We will execute this operation regularly and set to vacuum (0) .I know this statement will clean up the data of the old version, but keep the latest version.We do not specify a version for the query data, and use the latest version by default.
> I also have a purge job (deletedFileRetentionDuration is 0 in my delta table) which will running in parallel. The purge job run a delete command first to remove the data genrated 7 days ago, then run vacuum command.

vacuum 0 or deletedFileRetentionDuration=0 is not safe as vacuum may delete parquet files not committed yet, and cause the issue mentioned here.

We document the caution of misconfiguration for vacuum [here](https://docs.delta.io/latest/delta-utility.html#remove-files-no-longer-referenced-by-a-delta-table):

> It is recommended that you set a retention interval to be at least 7 days, because old snapshots and uncommitted files can still be in use by concurrent readers or writers to the table. If VACUUM cleans up active files, concurrent readers can fail or, worse, tables can be corrupted when VACUUM deletes files that have not yet been committed. You must choose an interval that is longer than the longest running concurrent transaction and the longest period that any stream can lag behind the most recent update to the table.

**Comment by ChenXiongfeng** [comment]
I also found this issue. It is really my misuse. thank you @zsxwing
